src/dbnode/storage/index/aggregate_results.go

```diff
@@ -101,201 +95,114 @@ func (r *aggregatedResults) Reset(
 	// reset all keys in the map next
 	r.resultsMap.Reset()
 	r.totalDocsCount = 0
+	r.size = 0
 
 	// NB: could do keys+value in one step but I'm trying to avoid
 	// using an internal method of a code-gen'd type.
 	r.Unlock()
 }
 
-func (r *aggregatedResults) AddDocuments(batch []doc.Document) (int, int, error) {
-	r.Lock()
-	err := r.addDocumentsBatchWithLock(batch)
-	size := r.resultsMap.Len()
-	docsCount := r.totalDocsCount + len(batch)
-	r.totalDocsCount = docsCount
-	r.Unlock()
-	return size, docsCount, err
-}
-
 func (r *aggregatedResults) AggregateResultsOptions() AggregateResultsOptions {
 	return r.aggregateOpts
 }
 
 func (r *aggregatedResults) AddFields(batch []AggregateResultsEntry) (int, int) {
 	r.Lock()
-	valueInsertions := 0
+	defer r.Unlock()
+
+	maxDocs := int(math.MaxInt64)
+	if r.aggregateOpts.DocsLimit != 0 {
+		maxDocs = r.aggregateOpts.DocsLimit - r.totalDocsCount
+	}
+
+	// NB: already hit doc limit.
+	if maxDocs <= 0 {
+		for idx := 0; idx < len(batch); idx++ {
+			batch[idx].Field.Finalize()
+			for _, term := range batch[idx].Terms {
+				term.Finalize()
+			}
+		}
+
+		return r.size, r.totalDocsCount
+	}
+
+	// NB: cannot insert more than max docs, so that acts as the upper bound here.
+	maxInserts := maxDocs
+	if r.aggregateOpts.SizeLimit != 0 {
+		if remaining := r.aggregateOpts.SizeLimit - r.size; remaining < maxInserts {
+			maxInserts = remaining
+		}
+	}
+
+	limitTripped := false
+	docs := 0
+	numInserts := 0
 	for _, entry := range batch {
+		if docs >= maxDocs || numInserts >= maxInserts {
+			limitTripped = true
+		}
+
+		if limitTripped {
```

Comment:
can we in-line the check above and remove the `limitTripped` var?

Reply:
Good call